### ISSUE TITLE
Version 1.0.2 of Java driver not thread safe

### ISSUE THREAD

**Issue by graemerocher** [neo4j/neo4j-java-driver#184]
Unless I am missing something the driver doesn't appear to be thread safe, which makes it unusable for most production scenarios. When accessed via multiple threads frequent exceptions occur:

```
org.neo4j.driver.v1.exceptions.ClientException: You are using a session from multiple locations at the same time, which is not supported. If you want to use multiple threads, you should ensure that each session is used by only one thread at a time. One way to do that is to give each thread its own dedicated session.

    at org.neo4j.driver.internal.connector.ConcurrencyGuardingConnection.markAsInUse(ConcurrencyGuardingConnection.java:192)
    at org.neo4j.driver.internal.connector.ConcurrencyGuardingConnection.run(ConcurrencyGuardingConnection.java:65)
    at org.neo4j.driver.internal.pool.PooledConnection.run(PooledConnection.java:64)
    at 
```

Initially I thought this was a bug in our object mapping implementation, but I wrote a test that uses the raw driver and the same issue occurs:

https://github.com/grails/grails-data-mapping/blob/master/grails-datastore-gorm-neo4j/src/test/groovy/grails/gorm/tests/MiscSpec.groovy#L156

**Comment by pontusmelke** [comment]
You shouldn't share a session between multiple threads, the driver maintains a pool of sessions so each thread should just ask the driver for a session and then close it when you are done with it.

Regards,
Pontus

**Comment by graemerocher** [comment]
So that is what I'm doing, see the code:

``` groovy
(1..count).eachParallel { counter ->
    def session = boltDriver.session()
    def tx = session.beginTransaction()

    tx.run("CREATE (n1:Team {props})", [props:[name:"Team $count".toString()]])
    tx.success()
    session.close()
}
```

Each closure is executed in a separate thread. I ask the driver for a session for the current thread. I then start a transaction, run CREATE and close the session.

I don't see where I am sharing a session amongst threads in this example?

**Comment by pontusmelke** [comment]
Oh sorry, that doesn't seem right. Are you using the latest version of the driver?

We'll investigate and release a fix for this issue as soon as possible.

**Comment by graemerocher** [comment]
I'm using

``` groovy
compile 'org.neo4j.driver:neo4j-java-driver:1.0.2'
```

Thanks

**Comment by pontusmelke** [comment]
You could try to use 1.0.1 until the fix has been rolled out.

**Comment by pontusmelke** [comment]
Oh and by the way I think you should also call `tx.close()` before `session.close()`.

The usage pattern (in java) is

```
try(Session session = driver.session()) {
    try(Transaction tx = session.driver()) {
    .... do stuff with the transaction
    }
}
```

**Comment by graemerocher** [comment]
I'll try with 1.0.1... I assume that the session would close the transaction, from https://github.com/neo4j/neo4j-java-driver/blob/1.1/driver/src/main/java/org/neo4j/driver/internal/InternalSession.java#L120

but anyway this is just a test and adding `close()` to the transaction doesn't change the outcome

**Comment by graemerocher** [comment]
I can confirm the issue is not present in the 1.0.1 version of the driver

**Comment by pontusmelke** [comment]
We have released a version 1.0.3 of the java driver which should fix the issue. Thanks for letting us know!

**Comment by graemerocher** [comment]
Great thanks.
